### PR TITLE
Pause Polls on VxScan

### LINKED ISSUE #2683: Allow Polls Suspension on VxScan
See the [Tech Spec](https://docs.google.com/document/d/15mSk4QPQUqsLlHFwF5Dw058OsprmoJixwxm7C2GmGGk/edit?usp=sharing) for the overall [roadmap item](https://github.com/votingworks/vxsuite-roadmap/issues/37).

Implement polls suspension on VxScan, including:
- Creating Polls State type(s)
- Managing polls state on the "backend"
- Updating poll worker screens
- Updating idle screens

This chunk of work will not update the printed reports or reports saved to the tally, which is tracked in a separate ticket.

### PR DESCRIPTION
**Pull request by adghayes**
## Overview
Closes #2683. Changes two main things about the election flow:
1. After polls are closed, polls cannot be re-opened. Poll Workers will not be able to re-open the polls and Election Managers will not be able to re-open the polls without reconfiguring first. System Administrators have an option to reset polls from closed to paused.
2. Adds polls paused (or "suspended" according to VVSG language) state. In this state, voting cannot take place but polls _can_ be re-opened.

We also place more limits on when election managers can change the precinct on a device. Formerly, they could change the precinct at any time. Now, they can change it freely while polls are not open. While polls are open but no ballots are not cast, they can still change the precinct but it will require a modal confirmation, reset the polls to unopened, and reprint the polls report.

Some things that are _not_ in this change:
- The new polls transition types (paused, unpaused) will _not_ be represented on the printed reports or in VxMark in any way. That will come in the next two PRs (VxMark and reports respectively).
- Testing for the previous bullet, since it doesn't exist yet

## Demo Video or Screenshot
Poll Worker Flow:
[all_states_in_flow.webm](https://user-images.githubusercontent.com/37960853/198367053-1c91aea3-db32-4786-8108-5abad0e252e2.webm)

Sys Admin Reset Flow:
[reset_polls_paused.webm](https://user-images.githubusercontent.com/37960853/198367073-9438d51c-2b4f-45a1-89ea-a00fc89b6f30.webm)

Election Manager Precinct Change Flow:
[change_precinct.webm](https://user-images.githubusercontent.com/37960853/198367157-b270dec1-37c4-473e-a49b-ce5ecc58cdb8.webm)

## Testing Plan
I refactored a lot of our testing in the frontend and wrote a good bit of new testing. Some backend and utility testing too. Manual testing.

## Checklist
- [x] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.
- [ ] I have added JSDoc comments to any newly introduced exports

- [x] I have added a screenshot and/or video to this PR to demo the change
- [x] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

File: libs/utils/src/precinct_selection.ts
Comment (adghayes) [inline libs/utils/src/precinct_selection.ts:49]: Not sure if there's a more idiomatic name I should use for this type of function, like `compare`.

File: libs/api/src/services/scan/index.ts
Comment (adghayes) [inline libs/api/src/services/scan/index.ts:326]: Yes please, I'll have two orders of boilerplate.

File: services/scan/src/precinct_scanner_app.ts
Comment (adghayes) [inline services/scan/src/precinct_scanner_app.ts:615]: I'm not sure if this is too limiting - to disabled scanning when polls are closed. Any reservations about this? Seems like a security in depth way to lock down "no voting while polls closed" requirements.

File: frontends/precinct-scanner/src/app_root.tsx
Comment (adghayes) [inline frontends/precinct-scanner/src/app_root.tsx:144]: We're not using this, just removed it.

File: frontends/precinct-scanner/src/app_root.tsx
Comment (adghayes) [inline frontends/precinct-scanner/src/app_root.tsx:394]: Because we may change other things, like polls state, when changing precinct selection, we need to refresh config after.

File: frontends/precinct-scanner/src/screens/poll_worker_screen.tsx
Comment (adghayes) [inline frontends/precinct-scanner/src/screens/poll_worker_screen.tsx:519]: @mattroe has been suggesting / asking / arguing to remove the export results option from the poll worker flow for a while, and I decided to do it here. I didn't want to update existing coverage, and we didn't have great coverage on it anyway.

File: frontends/precinct-scanner/src/app_root.tsx
Comment (carolinemodic) [inline frontends/precinct-scanner/src/app_root.tsx:356]: We should log whenever this updates. If you look at the table on page 253 of VVSG it gives some specific language about logging opening and closing of polls and also needing to log "The “zero totals” check conducted before opening
the polling place or counting a precinct centrally"

File: frontends/precinct-scanner/src/screens/election_manager_screen.tsx
Comment (carolinemodic) [inline frontends/precinct-scanner/src/screens/election_manager_screen.tsx:194]: Is it possible to reuse this component for the initial selection of a precinct as well? It feels like we are duplicating a lot in this component and in the code above here.

File: frontends/precinct-scanner/src/screens/poll_worker_screen.tsx
Comment (carolinemodic) [inline frontends/precinct-scanner/src/screens/poll_worker_screen.tsx:70]: I assume this is for a future PR?

File: frontends/precinct-scanner/src/screens/poll_worker_screen.tsx
Comment (carolinemodic) [reply frontends/precinct-scanner/src/screens/poll_worker_screen.tsx:519]: I'm not sure I understand the desire of removing it but I guess I defer to @mattroe

File: libs/types/src/polls.ts
Comment (carolinemodic) [reply libs/types/src/polls.ts:3]: I might be missing some decision that you all have made here about the best practice though.

File: libs/api/src/services/scan/index.ts
Comment (eventualbuddha) [reply libs/api/src/services/scan/index.ts:326]: All the boilerplate you could want:
image

File: frontends/precinct-scanner/src/app_root.tsx
Comment (carolinemodic) [inline frontends/precinct-scanner/src/app_root.tsx:461]: why do we have this boolean if its just always true?

File: frontends/precinct-scanner/test/helpers/mock_polls_state.ts
Comment (carolinemodic) [inline frontends/precinct-scanner/test/helpers/mock_polls_state.ts:4]: love these helpers

**Review by carolinemodic — CHANGES_REQUESTED** [review]

File: services/scan/src/precinct_scanner_app.ts
Comment (carolinemodic) [reply services/scan/src/precinct_scanner_app.ts:615]: I like this

File: frontends/precinct-scanner/src/app_root.tsx
Comment (adghayes) [reply frontends/precinct-scanner/src/app_root.tsx:461]: It's not true on VxCentralScan, where I think this screen is also used.

File: frontends/precinct-scanner/src/screens/poll_worker_screen.tsx
Comment (adghayes) [reply frontends/precinct-scanner/src/screens/poll_worker_screen.tsx:70]: Yup!

File: frontends/precinct-scanner/src/screens/election_manager_screen.tsx
Comment (adghayes) [reply frontends/precinct-scanner/src/screens/election_manager_screen.tsx:194]: Could be, I think we might want to reuse it in the BMD too. I will take a look.

File: frontends/precinct-scanner/src/app_root.tsx
Comment (adghayes) [reply frontends/precinct-scanner/src/app_root.tsx:356]: We are lacking _a lot_ of logging on precinct scanner and I held off on some of it because I know we'll have to do a general pass later. I think it may be cleaner to do all that (polls state change, precinct state change (your comment above), report printed, saved to card) in a separate PR?

File: frontends/precinct-scanner/src/app_root.tsx
Comment (adghayes) [reply frontends/precinct-scanner/src/app_root.tsx:394]: I don't think it's a bug right now, since right now nothing else can change when the precinct gets updated.

Comment (adghayes) [reply frontends/precinct-scanner/src/app_root.tsx:394]: See my other comment about logging, I'd like to do a logging pass in VxScan as a separate PR.

File: libs/types/src/polls.ts
Comment (adghayes) [reply libs/types/src/polls.ts:3]: You will get a Typescript error if you have a typo in the string. @arsalansufi  has been suggesting union types over enums for best practice, let me see if I can find the article...

File: frontends/precinct-scanner/src/screens/poll_worker_screen.tsx
Comment (adghayes) [reply frontends/precinct-scanner/src/screens/poll_worker_screen.tsx:519]: It's not a feature that we have a reasonable user story for.

File: services/scan/src/store.ts
Comment (jonahkagan) [inline services/scan/src/store.ts:75]: The polls state feels like a different class of data than the rest of the config values. Those are used to set up ("configure") the machine, while this is more a piece of operating state that has to do with using the machine (it's not a setting).

What do you think about storing it separately and using separate API routes? Does this seem too pedantic? The motivation would mostly just be clarity.

File: libs/types/src/polls.ts
Comment (adghayes) [reply libs/types/src/polls.ts:3]: Typescript error:
Screen Shot 2022-10-27 at 1 42 23 PM

Comment (adghayes) [reply libs/types/src/polls.ts:3]: https://github.com/votingworks/vxsuite/pull/2221#discussion_r932457382

File: services/scan/src/store.ts
Comment (adghayes) [reply services/scan/src/store.ts:75]: It does seem too pedantic in isolation, but I'm thinking about how as part of this project I plan to move the last bits of frontend state (ballot replaced count and `isSoundsMuted`). Since we're going to have more of these values, I think it makes sense to add the clarity of breaking it out. 

What do you think about all settings being under a single API endpoint that you update in a `Partial` fashion? Just trying to limit code, but maybe that's the wrong motivation.

Comment (jonahkagan) [reply services/scan/src/store.ts:75]: Yeah, updating settings with a single endpoint seems fine.

Here's my take on how I would categorize the different bits of state:
- "Config" refers to setting up the machine to work in the right context (e.g. election definition, templates, precinct selection) (we might want to rename this to avoid confusion with machine config - machine id, etc)
- "Settings" are user settings that modify machine behavior, but don't have to do with the election (e.g. muting)
- "Operating state" tracks what's happening while the machine is being used (e.g. test mode, polls state) (this category maybe needs a better name)

Does that seem intuitive/clear to you, or would you think about it differently?

File: frontends/precinct-scanner/src/screens/election_manager_screen.tsx
Comment (adghayes) [reply frontends/precinct-scanner/src/screens/election_manager_screen.tsx]: Refactored.

File: services/scan/src/store.ts
Comment (adghayes) [reply services/scan/src/store.ts:75]: This seems intuitive and clear to me, mostly. My only confusion is that, one way I've thought about config in the past is that it's something that affects how we interpret ballots. and "Test Mode" does affect how we read ballots. So on that piece, it's a big unclear. 

Given this is looking like a bigger change than I expected, I would prefer to do this in a different PR and merge the current polls open state as is, change later.

Comment (eventualbuddha) [reply services/scan/src/store.ts:75]: FWIW I want to kill the `configs` table and replace it with specific tables/columns/etc for the data we want to store.

File: frontends/precinct-scanner/src/components/change_precinct_button.tsx
Comment (adghayes) [inline frontends/precinct-scanner/src/components/change_precinct_button.tsx:155]: I added this `assertDefined` utility because I think it's valuable to be able to indicate, in code, that something should be defined, just like to do with `assert`. It is essentially re-introducing the bang operator, and `assert` is preferable where possible. But these cases, using the existence of a prop as a condition for rendering the element may create the false impression that the element is not meant to be rendered under those conditions.

In @jonahkagan's words, better to fail loudly. cc @arsalansufi who discussed this with me.

File: services/scan/src/store.ts
Comment (adghayes) [reply services/scan/src/store.ts:75]: @eventualbuddha where do you envision just one-off settings, like `isSoundMuted` living? All in their own column / table?

File: libs/types/src/polls.ts
Comment (carolinemodic) [reply libs/types/src/polls.ts:3]: Ah I see ok, I think typescript has improved how it deals with this since the last time I tried haha

File: frontends/precinct-scanner/src/app_root.tsx
Comment (carolinemodic) [reply frontends/precinct-scanner/src/app_root.tsx:356]: Yeah there are definitely a lot of other areas where we are missing logging, I think its probably more efficient in general to just add things when we're in that code as its a little hard later sometimes to remember exactly what the error cases are or whats happening, and doing all of the logging later is a task no one is going to want to do. But we don't have to block on it

File: services/scan/src/store.ts
Comment (eventualbuddha) [reply services/scan/src/store.ts:75]: Assuming we have an `elections` table similar to VxAdmin, these settings could either live on that table (if they're tied to an election) or in a `settings` table that's essentially a singleton table.

Comment (adghayes) [reply services/scan/src/store.ts:75]: I would prefer to separate out the reshuffling of how we store / access state from this PR. I propose I leave everything as-is for now, and I will revisit when I work on #2687, which is when I'll move the rest of the state back. Between now and then, we can sync for 15 min and sketch out how we'd like to lay out state?

File: frontends/precinct-scanner/src/components/change_precinct_button.tsx
Comment (eventualbuddha) [reply frontends/precinct-scanner/src/components/change_precinct_button.tsx:155]: I've written a utility like this at various times in my various TypeScript projects. It's better than a type cast/assertion (whether `as` or `!`) because it will fail _here_ and not at some later point where the value being used is not as expected. It's sometimes worse in terms of readability, but on the whole I think it's good.

Comment (adghayes) [reply frontends/precinct-scanner/src/components/change_precinct_button.tsx]: Sounds good. I ended up not needing it right here but have left it in this PR because I can anticipate needing it again.

File: frontends/precinct-scanner/src/app_root.tsx
Comment (carolinemodic) [reply frontends/precinct-scanner/src/app_root.tsx:461]: semi seperately should we also not show the button at all if the polls aren't in polls_closed_final?

Comment (adghayes) [reply frontends/precinct-scanner/src/app_root.tsx:461]: Right now it's just disabled if the polls aren't in `polls_closed_final`, I think that's fine for now. I actually originally had it such that it was hidden if polls were not in `polls_closed_final` but React / the testing didn't like how it was being removed the the page (trying to close the modal after the higher level state had already removed it from the page).

File: frontends/precinct-scanner/src/app_tally_report_paths.test.tsx
Comment (carolinemodic) [inline frontends/precinct-scanner/src/app_tally_report_paths.test.tsx:1366]: AFAICT we don't have any testing right now for the polls paused straight to poll closed transition, probably worth adding a printing and save to card test for that flow.

**Review by carolinemodic — APPROVED** [review]
One suggested test I thought off after watching the videos but otherwise this lgtm

File: frontends/precinct-scanner/src/app_tally_report_paths.test.tsx
Comment (adghayes) [reply frontends/precinct-scanner/src/app_tally_report_paths.test.tsx:1366]: Yeah, I skipped that that one since it's not a use case we expect very much. But, having a test is better :) Will add then merge.

**Comment by eventualbuddha** [comment]
From the first video:

> it doesn't indicate right on this screen that you are coming from a paused state since that should be clear from the previous screen

I wonder about using the phrase "Open Polls" to describe what's happening here. To me, opening the polls is something that happens _at the start of voting_. And it appears we don't get a polls opened report when we do this "Open Polls". That _feels_ weird to me. I'm imagining a poll worker thinking "I opened the polls this morning and got a report, but I opened the polls again in the afternoon and didn't get a report. Why not? Should I have?". I'd argue we should say more precisely what we're doing: "Open Polls" / "Pause Polls" / "Unpause Polls" (or "Resume Polls"?) / "Close Polls".

I of course defer to @mattroe and @ginvdr on this, just wanted to register my discomfort with that.

**Comment by adghayes** [comment]
@eventualbuddha I wondered the same thing, and we could definitely change it.

My thinking was that poll workers will rarely be doing this in the middle of the day. The main use case for this feature is a vote center model, where you'd start voting at the beginning of each day. If it's a new day of voting, "Open Polls" makes sense to me, even if polls were opened yesterday. The current plan is that the report will say "Polls Opened Report", but it will have an appearance different from the first day polls opened report. And it will say on the report that it was opened from a paused state.

Will wait for @mattroe and/or @ginvdr to weigh in.

**Comment by adghayes** [comment]
Larger context - there will be a report with every single polls state change. The one's for pausing and unpausing will only have ballot counts, not results, per VVSG.

**Comment by jonahkagan** [comment]
I had the same wondering about terminology as @eventualbuddha!

**Pull request merged**
